### ISSUE TITLE
Bug: MobX-like observer pattern doesn't work with Fast Refresh because Hooks don't get detected

### ISSUE THREAD

**Issue by Bnaya** [facebook/react#20417]
React version: 17.0.0

## Steps To Reproduce

1. https://codesandbox.io/s/react-refresh-webpack-plugin-rendered-more-hooks-than-during-the-previous-render-issue-ezcrz?file=/src/Comp.js
2. Delete one of the hooks there

Link to code example:
https://codesandbox.io/s/react-refresh-webpack-plugin-rendered-more-hooks-than-during-the-previous-render-issue-ezcrz?file=/src/Comp.js

## The current behavior
You get "Rendered more hooks than during the previous render" error

## The expected behavior
Should hot reload and re-mount the component.

The source of the issue have two parts:
1. react-refresh and the bundler fails to inject signature to the component
2. When no signature apparent, `react-refresh` consider the components as compatible, which is not always true, as in the repro https://github.com/facebook/react/blob/9aca239f11f31109dc1a229aa1571c2bf02f5524/packages/react-refresh/src/ReactFreshRuntime.js#L126-L132

I've filed an issue also for the webpack plugin: https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/266
Mobx related issue: https://github.com/mobxjs/mobx/issues/2668

**Comment by Sadin1111** [comment]
> 
> 
> React version: 17.0.0
> 
> ## Steps To Reproduce
> 
> 1. https://codesandbox.io/s/react-refresh-webpack-plugin-rendered-more-hooks-than-during-the-previous-render-issue-ezcrz?file=/src/Comp.js
> 2. Delete one of the hooks there
> 
> 
> 
> Link to code example:
> https://codesandbox.io/s/react-refresh-webpack-plugin-rendered-more-hooks-than-during-the-previous-render-issue-ezcrz?file=/src/Comp.js
> 
> 
> 
> ## The current behavior
> You get "Rendered more hooks than during the previous render" error
> 
> 
> ## The expected behavior
> Should hot reload and re-mount the component.
> 
> The source of the issue have two parts:
> 1. react-refresh and the bundler fails to inject signature to the component
> 2. When no signature apparent, `react-refresh` consider the components as compatible, which is not always true, as in the repro https://github.com/facebook/react/blob/9aca239f11f31109dc1a229aa1571c2bf02f5524/packages/react-refresh/src/ReactFreshRuntime.js#L126-L132
> 
> I've filed an issue also for the webpack plugin: https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/266
> Mobx related issue: https://github.com/mobxjs/mobx/issues/2668

**Comment by Sadin1111** [comment]
5d65cbe3c8dca5f463613495003c5f96006a3555

**Comment by rizkyramadhan** [comment]
This is impacting user who uses mobx as their state management. As each of component should be wrapped with observer.

When component is wrapped in observer, fast-refresh does not work.

**Comment by TeosVerdi** [comment]
I'd love to find a solution (or at least workaround) for this issue.
It persists on Next.js without snowpack, only using react-refresh.
See https://github.com/mobxjs/mobx/discussions/2794#discussion-3151716

**Comment by rizkyramadhan** [comment]
Potential work around for mobx is to use `<Observer />` component instead of `observer  hoc, for example your old component:

```
const Component = observer(() => {
    return <div>Hello</div>
})
```

will become like this:

```
const Component = () => {
    return <Observer>{() => (
        <div>Hello</div>
    )}</Observer>
}
```

**Comment by TeosVerdi** [comment]
Sadly, it doesn't work in my case. I've tried arrow functions, Observer component and HOC: neither of them do the trick, the checkbox stays checked after refresh either way.

**Comment by AoDev** [comment]
@rizkyramadhan I have tried the <Observer/> approach but couldn't solve all the cases in the end.

**Comment by rizkyramadhan** [comment]
Yeah, it's pretty wierd... I ended up ditching snowpack altogether, Now I use customized esbuild for my bundler.

**Comment by yuhongda** [comment]
> Yeah, it's pretty wierd... I ended up ditching snowpack altogether, Now I use customized esbuild for my bundler.

Hi, can you talk more about the customized esbuild solution?

**Comment by AoDev** [comment]
@rizkyramadhan I dont use snowpack, nor do I know what it is. We have a "normal" setup with [react-refresh-webpack-plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin) without any sort of abstraction like CRA.

React Hot Loader V3 was the last version that worked properly for our mobx / react apps and we'd stayed on it if it were not for incompatibilities coming with new React versions. RHL v4 never properly worked with similar behaviour than react-refresh. Disappointing that we had a better developer experience 5 years ago.

**Comment by rizkyramadhan** [comment]
> > Yeah, it's pretty wierd... I ended up ditching snowpack altogether, Now I use customized esbuild for my bundler.
> 
> Hi, can you talk more about the customized esbuild solution?

Basically I build my own bundler, first I tested [simple esbuild solution using react](https://esbuild.github.io/getting-started/#your-first-bundle), and modify the esbuild parameters with various esbuild plugins for my needs, and setup chokidar to watch my src folders for changes. When source changes, I broadcast the changes using [esm-hmr protocol](https://github.com/snowpackjs/esm-hmr), currently it's still using naive dom replacement without react-refresh. But it's really fast, even comparable to react-refresh due to esbuild. 

The most important thing is mobx and another npm package that I use works out of the box.

**Comment by gaearon** [comment]
>react-refresh and the bundler fails to inject signature to the component

Have you looked into why?

**Comment by gaearon** [comment]
I think you want to look at these tests:

https://github.com/facebook/react/blob/7b84dbd1697609d4df5931422c9591fe62022cb9/packages/react-refresh/src/__tests__/ReactFreshBabelPlugin-test.js#L209-L222

Add something like the example in your repro. What does it get transformed to? Why doesn't it get the same treatment like `memo`?

The relevant logic is somewhere around here, probably:

https://github.com/facebook/react/blob/7b84dbd1697609d4df5931422c9591fe62022cb9/packages/react-refresh/src/ReactFreshBabelPlugin.js#L632-L640

**Comment by gaearon** [comment]
There seems to be a misunderstanding in this thread.

First, the original issue reported in https://github.com/facebook/react/issues/20417#issue-760661008 no longer reproduces on CodeSandbox. I think that this was a bug in how CodeSandbox implemented Fast Refresh, which they fixed in https://github.com/codesandbox/codesandbox-client/pull/5442. So it was fixed in February.

This is confirmed by https://github.com/facebook/react/issues/20417#issuecomment-754126769:

> this problem only affect snowpack project, CRA are fine.

Then, the comment in https://github.com/facebook/react/issues/20417#issuecomment-776204555 has nothing to do with that problem. The linked thread is about some checkbox not working, which is not at all the symptom described in this thread. So this was incorrectly lumped together with this issue. That issue is likely just something being broken in MobX.

Then, https://github.com/facebook/react/issues/20417#issuecomment-804683995 is a new comment about some problem but it doesn't specify what the problem actually is, or offer any reproducing case. So I can't help here and I can't guess which of the problems above it refers to.

In conclusion:

* I'm going to close this issue since the original problem was solved by CodeSandbox/Snowpack.
* https://github.com/mobxjs/mobx/discussions/2794#discussion-3151716 seems unrelated and there is no reason to think it has something to do with React so far.
* https://github.com/facebook/react/issues/20417#issuecomment-804683995 does not explain what the problem is, but if there is still a problem that isn't in MobX, please file a new issue with a repro case.

Thanks!

**Comment by gaearon** [comment]
After trying more, I can repro this now. So my conclusion seems incorrect. I'll take a closer look at this...

**Comment by gaearon** [comment]
Yayyy I can repro this in CRA. This is a good sign the problem is on our side.

**Comment by gaearon** [comment]
Hmm. So the problem here seems to be is that the `observer` thing doesn't actually *render* the component passed to it. It calls it *as a function*. So even though React knows that inner component needs to be remounted, the inner component is _nowhere in the tree_. Cause it's called as a function instead of rendered as a component.

**Comment by gaearon** [comment]
This is a pretty exotic and unidiomatic pattern tbh. But it makes sense that MobX does it since it needs to know what has been read.

**Comment by gaearon** [comment]
Let me describe the problem a bit clearer. Fast Refresh works by tagging functions with "signatures".

```js
// V1
function Foo() {
  useA()
}

// V2
function Foo() {
  useA()
  useB()
}
```

These two versions of `Foo` would get different "signatures", so we'd know to remount them.

These signatures are specified through calls to a special function.

```js
// V1
let Foo = _s(function Foo() {
  useA()  
}, 'abcde')

// V2
let Foo = _s(function Foo() {
  useA()  
  useB()  
}, 'qwerty')
```

The tricky part is what happens when it's a HOC. Because the part getting signed is the inner one.

```js
// V1
let Foo = observer(_s(function Foo() {
  useA()  
}, 'abcde'))

// V2
let Foo = observer(_s(function Foo() {
  useA()  
  useB()  
}, 'qwerty'))
```

This would normally work fine. React would see that the signature associated `Foo` is different, so all `Foo` components need to b e remounted.

The problem with what MobX is doing is that `Foo` does not end up being a component in the tree. MobX just calls it as a function: `Foo()`. The only component React sees is the MobX generated wrapper. React has no idea that the `_s` call it did for `Foo` actually describes the signature of the wrapper that calls `Foo()`.

React doesn't find `Foo` in the tree, and it thinks that MobX wrappers are completely safe to keep mounted, since their own signatures haven't changed.

**Comment by gaearon** [comment]
So I have a fix for this but I really don't like it. https://github.com/facebook/react/pull/21104

We'll probably fix MobX and break something else. I'll need to think more about this. This is pretty frustrating because MobX has significant usage but the pattern is using is very exotic and it feels bad to make concessions because of it.

**Comment by AoDev** [comment]
React Hot Loader v3 (not v4) worked fine for us despite reading here and there that react-refresh is more reliable.

I also understand react-refresh is expecting a "correct" way to use React.

A possibility could be to leave react-refresh "pure" and have a new package based on RHL v3 that would work. But we moved to react-refresh because it seemed that RHL v3 would not be compatible with future versions of React. (already getting various warnings)

**Comment by gaearon** [comment]
I get the frustration. I hope you can believe me as someone who spent significant time on this problem that the current approach really is the most stable one and handles many cases that RHL v3 would not be able to. However, here concretely are running into a very particular case we never intended to support. We can probably figure out some fix here but it would be ideal if MobX fixed it on their end instead, I think. I'll talk to Michel about this.

**Comment by AoDev** [comment]
Oh yes, I believe you. We have noted that mobx and React have different paths. For us React components look more like simple templates when using Mobx. That being said, your work during all these years has made our developer life so much better that a simple "thank you" will never be enough to show our appreciation. Thank you for everything : )

**Comment by gaearon** [comment]
Can you give this version a spin?

[react-refresh.zip](https://github.com/facebook/react/files/6209445/react-refresh.zip)

**Comment by yuhongda** [comment]
> Can you give this version a spin?
> 
> [react-refresh.zip](https://github.com/facebook/react/files/6209445/react-refresh.zip)
Just give it a try, but it's just like before :(

"react": "^17.0.1",
"mobx": "^6.1.5",
"mobx-react-lite": "^3.2.0",
"snowpack": "^3.0.13",
"@snowpack/plugin-react-refresh": "^2.3.8", (replaced the react-refresh package)

**Comment by gaearon** [comment]
Do you have a minimal repro? I used the one in the beginning of this thread and it works for me locally. But it’s using a reduced case instead of actual MobX.

**Comment by AoDev** [comment]
I have tried the version you shared. It didn't fix the issue.

I am using [mmmwh/react-refresh-webpack-plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin).

I just replaced  react-refresh in my node_modules.
Here is the repo I am testing it with: https://github.com/AoDev/bard-app.

(It's a boilerplate I use for my projects.)

```
npm install
npm run dev
```

My test consists of updating the SignIn form (first thing on screen) by adding / removing some text in it.

https://github.com/AoDev/bard-app/blob/master/src/application/App/Public/SignIn/SignIn.js#L19

![image](https://user-images.githubusercontent.com/1526150/112646168-84fe1b80-8e47-11eb-9c78-a5e33abd76f7.png)

Getting this in logs

```
[HMR] bundle rebuilding
client.js?146e:250 [HMR] bundle rebuilt in 537ms
process-update.js?881f:51 [HMR] Checking for updates on the server...
process-update.js?881f:125 [HMR] Updated modules:
process-update.js?881f:127 [HMR]  - ./src/application/App/Public/SignIn/SignIn.js
process-update.js?881f:132 [HMR] App is up to date.
```

But no refresh.

**Comment by yuhongda** [comment]
> Do you have a minimal repro? I used the one in the beginning of this thread and it works for me locally. But it’s using a reduced case instead of actual MobX.

sure thing. please check it. [https://github.com/yuhongda/repro-react-issues-20417](url)

**Comment by gaearon** [comment]
>But no refresh.

To be clear, _this_ issue was about an error being thrown ("rendered too many Hooks"). That's what my fix is supposed to help with. Are you describing a different problem?

**Comment by gaearon** [comment]
>sure thing. please check it. https://github.com/yuhongda/repro-react-issues-20417

What are the reproducing steps?

**Comment by AoDev** [comment]
> > But no refresh.
> 
> To be clear, _this_ issue was about an error being thrown ("rendered too many Hooks"). That's what my fix is supposed to help with. Are you describing a different problem?

I am coming from a serie of issues reported by various mobx users who don't see any update.

I don't know the underlying cause and thought this one was related as it was pointed to me [here](https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/301).

It is possible that the issue described here is not the cause but I thought it was.

**Comment by gaearon** [comment]
Ah I see. That's probably a separate issue. Although there might be related underlying causes. But I haven't looked at that one yet.

**Comment by yuhongda** [comment]
> > sure thing. please check it. https://github.com/yuhongda/repro-react-issues-20417
> 
> What are the reproducing steps?

Please follow the steps:

1. install & run it
`
yarn
yarn dev
`
2. and then snowpack will open a browser tab automatically, just like below:
![image](https://user-images.githubusercontent.com/259410/112661245-47be7b80-8e91-11eb-9780-7eda45b2f643.png)

3. find `./src/components/Home/index.tsx`, edit & save
![image](https://user-images.githubusercontent.com/259410/112661775-d8955700-8e91-11eb-8ddc-c50a43ae2566.png)

4. check the logs
![image](https://user-images.githubusercontent.com/259410/112661933-0d091300-8e92-11eb-8745-d048ca3a0bb4.png)

5. but the page not get updated.

**Comment by yuhongda** [comment]
> > sure thing. please check it. https://github.com/yuhongda/repro-react-issues-20417
> 
> What are the reproducing steps?



> I have tried the version you shared. It didn't fix the issue.
> 
> I am using [mmmwh/react-refresh-webpack-plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin).
> 
> I just replaced react-refresh in my node_modules.
> Here is the repo I am testing it with: https://github.com/AoDev/bard-app.
> 
> (It's a boilerplate I use for my projects.)
> 
> ```
> npm install
> npm run dev
> ```
> 
> My test consists of updating the SignIn form (first thing on screen) by adding / removing some text in it.
> 
> https://github.com/AoDev/bard-app/blob/master/src/application/App/Public/SignIn/SignIn.js#L19
> 
> ![image](https://user-images.githubusercontent.com/1526150/112646168-84fe1b80-8e47-11eb-9c78-a5e33abd76f7.png)
> 
> Getting this in logs
> 
> ```
> [HMR] bundle rebuilding
> client.js?146e:250 [HMR] bundle rebuilt in 537ms
> process-update.js?881f:51 [HMR] Checking for updates on the server...
> process-update.js?881f:125 [HMR] Updated modules:
> process-update.js?881f:127 [HMR]  - ./src/application/App/Public/SignIn/SignIn.js
> process-update.js?881f:132 [HMR] App is up to date.
> ```
> 
> But no refresh.

I just think maybe we have the same problem.

**Comment by gaearon** [comment]
Pushed another commit to the fix. It solves the crash in https://github.com/mobxjs/mobx/issues/2668#issue-757517589 (my original fix wasn't enough). It doesn't solve the issue in the other comments though. I don't think these are the same issues so I'll keep looking.

[react-refresh.zip](https://github.com/facebook/react/files/6213160/react-refresh.zip)

**Comment by gaearon** [comment]
Regarding @AoDev's issue. It's MobX idioms strike again, unfortunately.

You have a component like this:

```js
export default function SignIn(props) {
  // ...
}
```

This component is fine. Here is how it's used:

```js
import withVM from 'bard-instruments/lib/mobx/withVM'
import SignIn from './SignIn'
import SignInVM from './SignInVM'
export default withVM(SignIn, SignInVM)
```

What is `withVM`? It's a MobX helper that again **does not render your component, instead calling it directly**.

This again breaks the expectations of React that things that look like components, and are exported as component, are, in fact, components. React thinks that it can just find `SignIn` in the tree and update it. But `SignIn` is nowhere to be found in the tree because in practice it's a MobX wrapper created in another file that calls your component as if it was a regular function.

I can fix it by doing this:

```js
export default function Blabla(props) {
  return <SignIn {...props} />
}

function SignIn(props) {
  // ...
}
```

At least with the most recent version I posted. But I don't really think there's something we can do to support this pattern. If you export a *component*, you need to *render* it, not call it. Either MobX needs to change its idioms to make this work, or you can export the MobX wrapper in the same file. I verified that this works.

**Comment by yuhongda** [comment]
> Pushed another commit to the fix. It solves the crash in [mobxjs/mobx#2668 (comment)](https://github.com/mobxjs/mobx/issues/2668#issue-757517589) (my original fix wasn't enough). It doesn't solve the issue in the other comments though. I don't think these are the same issues so I'll keep looking.
> 
> [react-refresh.zip](https://github.com/facebook/react/files/6213160/react-refresh.zip)

This one you shared is also not work :(

**Comment by gaearon** [comment]
@yuhongda I can't run your example because it uses an npm mirror. If I remove the mirror I get the error that `"https://registry.yarnpkg.com/@scf%2freact-templates` is not found.

**Comment by yuhongda** [comment]
> @yuhongda I can't run your example because it uses an npm mirror. If I remove the mirror I get the error that `"https://registry.yarnpkg.com/@scf%2freact-templates` is not found.

I just update it. Please try it again.

**Comment by gaearon** [comment]
@yuhongda Now getting this.

<img width="1055" alt="Screenshot 2021-03-26 at 17 32 32" src="https://user-images.githubusercontent.com/810438/112670793-47a28980-8e59-11eb-8165-c7c295b3d9ad.png">

**Comment by gaearon** [comment]
And then if I remove the router store altogether, I get this. Something is pretty broken in this setup.

<img width="1011" alt="Screenshot 2021-03-26 at 17 34 14" src="https://user-images.githubusercontent.com/810438/112670985-87697100-8e59-11eb-87bc-d2b9bc598962.png">

**Comment by yuhongda** [comment]
> @yuhongda Now getting this.
> 
> <img alt="Screenshot 2021-03-26 at 17 32 32" width="1055" src="https://user-images.githubusercontent.com/810438/112670793-47a28980-8e59-11eb-8165-c7c295b3d9ad.png">

please check if the snowpack version == 3.0.13

**Comment by gaearon** [comment]
OK that runs yea.

**Comment by yuhongda** [comment]
> OK that runs yea.

great!

**Comment by gaearon** [comment]
@yuhongda The issue you're experiencing is a problem with your bundler or some transformation. The code of `node_modules/react-refresh/cjs/react-refresh-runtime.development.js` has these lines:

```js
    if (typeof type === 'object' && type !== null) {
      switch (getProperty(type, '$$typeof')) {
```

But the code I see in the browser is:

```js
    if (typeof type === 'object' && type !== null) {
      switch (getProperty(type, '$typeof')) { // NOTE: ONLY ONE $
```

This is why my fix does not work for you.

**Comment by gaearon** [comment]
To sum up:

- The original problem in this issue would be solved by https://github.com/facebook/react/pull/21104 (if we proceed).
- The problem @AoDev is seeing is an issue of exporting a component but later *calling* it from another file. The fix is to either change MobX to stop relying on unidiomatic patterns (unlikely) or to move `withVM` call into the same file as the component definition.
- The problem @yuhongda is seeing is a bug in snowpack or some other transform step which mangles the source code.

**Comment by yuhongda** [comment]
> @yuhongda The issue you're experiencing is a problem with your bundler or some transformation. The code of `node_modules/react-refresh/cjs/react-refresh-runtime.development.js` has these lines:
> 
> ```js
>     if (typeof type === 'object' && type !== null) {
>       switch (getProperty(type, '$$typeof')) {
> ```
> 
> But the code I see in the browser is:
> 
> ```js
>     if (typeof type === 'object' && type !== null) {
>       switch (getProperty(type, '$typeof')) { // NOTE: ONLY ONE $
> ```
> 
> This is why my fix does not work for you.

I remove all the Babel plugins, the problem is still there. Maybe it’s problem about snowpack, and I’ll try it later.

**Comment by gaearon** [comment]
Clarification: I misunderstood @AoDev's repro case thinking that `useVM` is a MobX API. Instead, it appears to be a custom utility. Regardless, my recommendation here is to always create observers around the components you use explicitly, in the same file. So instead of exporting `SignIn` and later wrapping it in `useVM` that calls `observer` in another file, make `SignIn` itself export an `observer(() => ....)` call. That will fix the issue.

**Comment by AoDev** [comment]
The code for withVM can be seen here: [withVM](https://github.com/AoDev/bard-instruments/blob/master/src/mobx/withVM.tsx)

It is a helper I've used for years now, just changed a bit when React switched to hooks. It was created as a reusable solution based on one convention, which is that there is a RootStore available in context.

The point of doing it this way is the following: 
- complete separation of concerns: the react component is "pure" and can be tested without any interference of Mobx.
- The "VM" is just a localstore that gets automatically connected to context via Mobx utility. This one is also testeable as a pure reactive model, no React.

withVM just "connects" them.

**Comment by gaearon** [comment]
I get the motivation for why you'd want to structure it this way, but there is some tension here between the way React works, and the way MobX tries to make it work. Unfortunately, the issue is right at the middle of that tension.

React does not work reactively based on mutation. Like you noted, we have different paths. There's a number of things MobX can do to try to hide it away but we need to be clear that it'll never be fully in line with the paradigm.

Here's a concrete example that has nothing to do with Fast Refresh. Say you start here:

```js
export default function SignIn() {
  // read some mobx stuff
}
```

Then you refactor it like this:

```js
export default Wrapper() {
  return <SignIn />
}

function SignIn() {
  // read some mobx stuff
}
```

From React's point of view, this is a perfectly normal refactor. People do these literally every day and nothing breaks.

But I think the `withVM` API would subtly break if you did that. Because now the thing getting auto-wrapped into `observer` would be `Wrapper`. Since `SignIn` executes lazily (rather than eagerly during parent render, it would not be "observed" at all, and changes would fail to propagate.

This illustrates that it's a leaky abstraction, and there's only so much you can do to hide it away. So I would encourage to always wrap things in `observer` directly where it's used. This would solve the problem above _and also_ fix Fast Refresh for you. As well as other features that rely on composition.

If it's a problem for tests, you can mock MobX out. But something has to give here.

**Comment by AoDev** [comment]
Thanks for the feedback, I'll see if I can come up with a solution that can get the best of both worlds. To be honest I've been checking interesting projects like [mobx-jsx](https://github.com/ryansolid/mobx-jsx) but it's more an experiment than anything. It was already hard to get companies to use Mobx a couple of years ago, now asking them to not use React would be mission impossible ; )

**Comment by mweststrate** [comment]
@AoDev `observer` should always be applied straight at the component definition; this makes potential refactoring issues like the shown by @gaearon more explicit and easier to detect and it prevents potential mismatches between behavior of something that is called potentially wrapped and unwrapped in e.g. a unit test, or very tricky to spot, a recursive component call. But most importantly it makes it easier to verify the important "everything that renders observables should be marked observer" easier to verify. I'll make sure the docs on mobx side will be more explicit about this. The important realisation here is that `observer` is a higher order function, but not a higher order component. So your utility can take care of the injection, but not of the `observer` wrapping without breaking other things.

**Comment by AoDev** [comment]
I guess I have a misunderstanding of how observer works. 

Isn't  `export default observer(MyComponent)` the same as  
```
import MyComponent

export default observer(MyComponent)
```
Because that's what the utility does.

Note that this utility is mostly used for big app sections, like a root feature if you will, the other components are just plain observers.

**Comment by mweststrate** [comment]
Correct it's not the same, since 1) RHL works on a per file basis and 2) references to MyComponent might refer to the unwrapped component (for example in the `MyComponent` file itself, its tests, etc)

**Comment by AoDev** [comment]
Thank you both for your time. I've got my mobx-react recipes since `mobservable`. I probably missed something along the way. I'll move the observer decoration out and test.  : )

**Comment by gaearon** [comment]
I was always curious... what does “M” stand for? Michel? :-)

**Comment by mweststrate** [comment]
It was a bad word play: a mob of observables; MOBservable 😅. Since it is
so fine grained in its observation, there would be gazillion observables
more than when using e.g rxjs.

**Comment by yuhongda** [comment]
> To sum up:
> 
> * The problem @yuhongda is seeing is a bug in snowpack or some other transform step which mangles the source code.

@gaearon Eventually I find out where the problem happen. It's because `@snowpack/plugin-react-refresh` use `replace()` to insert react-refresh code, so the `$$` will be also replace by `$`. So that's the reason. I have commit pull request to fix it: 
[https://github.com/snowpackjs/snowpack/pull/3015](https://github.com/snowpackjs/snowpack/pull/3015)

Also. I need to update react-refresh version in dependencies of plugin-react-refresh. So can you give me the version tag of react-refresh with this fix?

Thanks a million :P

**Comment by Bnaya** [comment]
> Let me describe the problem a bit clearer. Fast Refresh works by tagging functions with "signatures".
> 
> ```js
> // V1
> function Foo() {
>   useA()
> }
> 
> // V2
> function Foo() {
>   useA()
>   useB()
> }
> ```
> 
> These two versions of `Foo` would get different "signatures", so we'd know to remount them.
> 
> These signatures are specified through calls to a special function.
> 
> ```js
> // V1
> let Foo = _s(function Foo() {
>   useA()  
> }, 'abcde')
> 
> // V2
> let Foo = _s(function Foo() {
>   useA()  
>   useB()  
> }, 'qwerty')
> ```
> 
> The tricky part is what happens when it's a HOC. Because the part getting signed is the inner one.
> 
> ```js
> // V1
> let Foo = observer(_s(function Foo() {
>   useA()  
> }, 'abcde'))
> 
> // V2
> let Foo = observer(_s(function Foo() {
>   useA()  
>   useB()  
> }, 'qwerty'))
> ```
> 
> This would normally work fine. React would see that the signature associated `Foo` is different, so all `Foo` components need to be remounted.
> 
> The problem with what MobX is doing is that `Foo` does not end up being a component in the tree. MobX just calls it as a function: `Foo()`. The only component React sees is the MobX generated wrapper. React has no idea that the `_s` call it did for `Foo` actually describes the signature of the wrapper that calls `Foo()`.
> 
> React doesn't find `Foo` in the tree, and it thinks that MobX wrappers are completely safe to keep mounted, since their own signatures haven't changed.

Sorry I'm late for the party, we are in a holiday:)
If im not mistaken, that issue can break any renderprop useage?
Is there a way to mark component as "always remount"?

**Comment by gaearon** [comment]
It don’t see how it would break render prop usage. I’d need a more specific example to explain why it doesn’t.

**Comment by gaearon** [comment]
`react-refresh@0.10.0` is out with the fix. I'll file an issue against the webpack plugin to update it, but feel free to upgrade it manually to try it in your project.

**Comment by eldh** [comment]
Fixes the problem for us, when used together with https://github.com/snowpackjs/snowpack/pull/3015. Thanks so much for fixing, Dan!

**Comment by Jack-Works** [comment]
If anyone using `react-refresh-typescript` (instead of `react-refresh/babel`) and having this problem, please cc me. Thanks!

**Comment by misaeldossantos** [comment]
Hi. I think i solved this problem inside react-refresh. I disabled the next update if the component name is "observerComponent". I don't know if I should do a PR as this is a mobx specific issue and maybe the react team doesn't want to include:

https://github.com/misaeldossantos/react/commit/f6eb19d8ccbab54b2daa6ec599371c8183d87fec.

Maybe it's possible to add a property in the component function to disable refresh in any component that needs it: Component.disableReactRefresh = true, but I'm not sure that's a good thing
